core/src/main/java/org/mule/runtime/core/message/MultiPartPayload.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.message;
+
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.toList;
+
+import org.mule.runtime.api.message.MuleMessage;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * Represents a payload of a {@link MuleMessage} composed of many different parts. Each parts is in itself a
+ * {@link MuleMessage}, and has {@code attributes} specific to that parts (such as http part headers).
+ * <p>
+ * This is useful for representing attachments as part of the payload of a message.
+ * 
+ * @since 4.0
+ */
+public class MultiPartPayload implements Serializable
+{
+    private static final long serialVersionUID = -1435622001805748221L;
+
+    /**
+     * The name of a part that does <b>not</b> represent an attachment.
+     */
+    public static final String BODY_PART_NAME = "_body";
```

Comment:
It would be best to not expose this, but rather expose a `Optional getBody()` to obtain body.